src/plugins/es_ui_shared/public/request/request.ts

```diff
@@ -36,121 +39,137 @@ export interface UseRequestConfig extends SendRequestConfig {
   deserializer?: (data: any) => any;
 }
 
-export interface UseRequestResponse {
+export interface UseRequestResponse<D = any, E = Error> {
   isInitialRequest: boolean;
   isLoading: boolean;
-  error: null | unknown;
-  data: any;
-  sendRequest: (...args: any[]) => Promise<SendRequestResponse>;
+  error: E | null;
+  data?: D | null;
+  sendRequest: (...args: any[]) => Promise<SendRequestResponse<D, E>>;
 }
 
-export const sendRequest = async (
-  httpClient: ng.IHttpService,
-  { path, method, body }: SendRequestConfig
-): Promise<SendRequestResponse> => {
+export const sendRequest = async <D = any, E = Error>(
+  httpClient: HttpSetup,
+  { path, method, body, query }: SendRequestConfig
+): Promise<SendRequestResponse<D, E>> => {
   try {
-    const response = await (httpClient as any)[method](path, body);
-
-    if (typeof response.data === 'undefined') {
-      throw new Error(response.statusText);
-    }
+    const stringifiedBody = typeof body === 'string' ? body : JSON.stringify(body);
+    const response = await httpClient[method](path, { body: stringifiedBody, query });
 
-    return { data: response.data, error: null };
+    return {
+      data: response.data ? response.data : response,
+      error: null,
+    };
   } catch (e) {
     return {
       data: null,
-      error: e.response ? e.response : e,
+      error: e.response && e.response.data ? e.response.data : e.body,
     };
   }
 };
 
-export const useRequest = (
-  httpClient: ng.IHttpService,
+export const useRequest = <D = any, E = Error>(
+  httpClient: HttpSetup,
   {
     path,
     method,
+    query,
     body,
     pollIntervalMs,
     initialData,
     deserializer = (data: any): any => data,
   }: UseRequestConfig
-): UseRequestResponse => {
+): UseRequestResponse<D, E> => {
+  const isMounted = useRef(true);
+  const sendRequestRef = useRef<() => Promise<SendRequestResponse<D, E>>>();
+  const scheduleRequestRef = useRef<() => void>();
```

Comment:
Can you explain why you created a ref here? I would have used `useCallback` for this purpose. Was it necessary to add the ref?

I see that `sendRequestRef` is also a ref. I think `useCallback` is better for this purpose. Maybe I'm missing something?

Reply:
I think switching to `useCallback` is an interesting idea and could yield optimization benefits by returning a memoized reference to `sendRequest`. It would look something like this, though I'm not sure about the implications of mutating a ref you're dependent upon (any risk of infinite loop?):

```ts
  const scheduleRequest = useCallback(
    () => {
      // Clear current interval
      if (pollIntervalIdRef.current) {
        clearTimeout(pollIntervalIdRef.current);
      }

      // Set new interval
      if (pollIntervalMsRef.current && isMounted.current) {
        pollIntervalIdRef.current = setTimeout(
          () => sendRequestRef.current!(),
          pollIntervalMsRef.current
        );
      }
    },
    [pollIntervalIdRef.current, pollIntervalMsRef.current, isMounted.current, pollIntervalIdRef.current, sendRequestRef.current],
  );
```

However, the original NP-ready code uses refs and I find them easier to reason about than the `useCallback` example above, so I think I'll leave optimization to another day.

Reply:
Yeah, it would imply a bigger refactor (e.g. why do we have a ref for `pollInterval`?), but I agree we can leave that to another day. 👍